Code/Framework/AzCore/AzCore/Jobs/JobExecutor.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/Jobs/Internal/JobTypeEraser.h>
+#include <AzCore/Jobs/JobDescriptor.h>
+#include <AzCore/std/containers/unordered_map.h>
+#include <AzCore/std/containers/vector.h>
+#include <AzCore/std/parallel/atomic.h>
+#include <AzCore/Memory/PoolAllocator.h>
+
+namespace AZ
+{
+    class JobGraphEvent;
+
+    namespace Internal
+    {
+        class CompiledJobGraph final
+        {
+        public:
+            AZ_CLASS_ALLOCATOR(CompiledJobGraph, SystemAllocator, 0)
+
+            CompiledJobGraph(
+                AZStd::vector<TypeErasedJob>&& jobs,
+                AZStd::unordered_map<uint32_t, AZStd::vector<uint32_t>>& links,
+                size_t linkCount,
+                bool retained);
+
+            ~CompiledJobGraph();
+
+            AZStd::vector<TypeErasedJob>& Jobs() noexcept
+            {
+                return m_jobs;
+            }
+
+            // Indicate that a constituent job has finished and decrement a counter to determine if the
+            // graph should be freed
+            void Release();
+
+        private:
+            friend class JobGraph;
+            friend class JobWorker;
+
+            AZStd::vector<TypeErasedJob> m_jobs;
+            AZStd::vector<uint32_t> m_successors;
+            AZStd::atomic<uint32_t>* m_dependencyCounts = nullptr;
```

Comment:
👍 The structure here is actually largely because I was spending (too much) time trying to get the type erasure working properly so a lot of logic had moved out of the job itself. This needs to be made cleaner I agree (as well as a more dedicated perf pass). Luckily, this is already performing better than what we already have so.. haha

Reply:
I can't easily bring the successor array into type erased job since this size scales with the number of edges in the graph, not the number of nodes. However, I changed it to store pointers instead if indices (removes one indirection) and moved the dependency counts into the tasks (as was originally intended in this case). The allocation of the successor array should be amortized for the majority case where the graph is retained, and on task completion, we'll just have to eat up the single cache fetch (+ hardware prefetch) to access all successors.

A downstream task can streamline this by pooling the jobs and successor pointers in thread local arenas.